scheduler/src/interfaces/agent_interface.rs

```diff
@@ -0,0 +1,118 @@
+    pub fn new() -> Self {
+        Self {
+            agent_pool: Arc::new(Mutex::new(AgentPool::new())), // Initialize the Agent Pool. It is lost when the Scheduler dies.
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl Agent for AgentService {
+    async fn register_agent(
+        &self,
+        request: tonic::Request<proto::Health>,
+    ) -> Result<tonic::Response<proto::RegisterAgentResponse>, tonic::Status> {
+        let input = request.get_ref();
+
+        info!("Received request from Agent: {:?}", input);
+        info!("\n  - Agent CPU usage: {}\n  - Agent memory usage: {}", input.cpu_avail, input.memory_avail);
+
+        // Lock the agent pool (to ensure thread-safe access) and handle potential mutex poisoning (instead of simply unwrapping and panicking)
+        let mut pool = match self.agent_pool.lock() {  // lock() returns the mutex content (guard, containing pool)
+            Ok(pool) => pool,
+            Err(poisoned) => {
+                error!("Agent pool lock poisoned, recovering...");
+                poisoned.into_inner()
+            }
+        };
+
+        let id = pool.generate_unique_id();
+        let score = compute_score(input.cpu_avail, input.memory_avail);
+
+        // Create a new Agent and add it to the Pool (it gets sorted)
+        let new_agent = PoolAgent::new(id, score);
+
+        // Response is the newly created Agent's ID.
+        let response = proto::RegisterAgentResponse { id: new_agent.get_id() };
+
+        pool.push(new_agent);
+
+        Ok(tonic::Response::new(response))
+    }
+
+    async fn report_health_status(
+        &self,
+        request: tonic::Request<tonic::Streaming<proto::HealthStatus>>,
+    ) -> Result<tonic::Response<proto::Empty>, tonic::Status> {
+        let mut stream = request.into_inner();
+
+        while let Some(health_status) = stream.next().await {
+            let status = match health_status {
+                Ok(status) => status,
+                Err(e) => {
+                    error!("Error receiving health status: {:?}", e);
+                    return Err(tonic::Status::internal("Error receiving health status"));
+                }
+            };
+
+            let health = match status.health {
+                Some(health) => health,
+                None => {
+                    error!("Health field is missing for Agent {}", status.agent_id);
+                    continue;  // Skip to the next health status if health data is missing
+                }
+            };
+
+            info!("Received health status from agent {}: CPU: {}, Memory: {}",
+                status.agent_id,
+                health.cpu_avail,
+                health.memory_avail
+            );
+
+            // Lock the agent pool (to ensure thread-safe access) and handle potential mutex poisoning (instead of simply unwrapping and panicking)
+            let mut pool = match self.agent_pool.lock() {  // lock() returns the mutex content (guard, containing pool)
```

Comment:
You can simply unwrap this

Reply:
Thing is, if i simply unwrap like:
```rust
let mut pool = self.agent_pool.lock().unwrap();
```

the code would panic at poisoning, which would kill the Scheduler

Here, the code
```rust
// Lock the agent pool (to ensure thread-safe access) and handle potential mutex poisoning (instead of simply unwrapping and panicking)
let mut pool = match self.agent_pool.lock() {  // lock() returns the mutex content (guard, containing pool)
    Ok(pool) => pool,
    Err(poisoned) => {
        error!("Agent pool lock poisoned, recovering...");
        poisoned.into_inner()
    }
};

```
handles mutex poisoning by recovering the lock and logging the error, without panicking

Which means the Scheduler server cannot be killed by a laser attack focused on its host machine to flip the bits in the part of the memory that holds the running process.

Reply:
Funny, but the unwrap on the lock is necessary. I am sorry :)

Reply:
Done